src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncIteratorRewriter.cs

```diff
@@ -339,51 +442,87 @@ private void GenerateIValueTaskSourceImplementation_OnCompleted()
             {
-                // Produce the implementation of `ValueTask IAsyncDisposable.DisposeAsync()`:
-                // this.builder.SetResult();
-                // this._valueOrEndPromise.Reset();
-                // this._state = StateNotStarted;
-                // return default;
+                // Produce an implementation for `void IValueTaskSource.GetResult(short token)`:
+                //  this._valueOrEndPromise.GetResult(token);
+                //  return;
 
-                NamedTypeSymbol IAsyncDisposable =
-                    F.WellKnownType(WellKnownType.System_IAsyncDisposable);
+                NamedTypeSymbol IValueTaskSource = F.WellKnownType(WellKnownType.System_Threading_Tasks_Sources_IValueTaskSource);
 
-                MethodSymbol IAsyncDisposable_DisposeAsync =
-                    F.WellKnownMethod(WellKnownMember.System_IAsyncDisposable__DisposeAsync)
-                    .AsMember(IAsyncDisposable);
+                MethodSymbol IValueTaskSource_GetResult =
+                    F.WellKnownMethod(WellKnownMember.System_Threading_Tasks_Sources_IValueTaskSource__GetResult)
+                    .AsMember(IValueTaskSource);
+
+                MethodSymbol promise_GetResult =
+                    F.WellKnownMethod(WellKnownMember.System_Threading_Tasks_Sources_ManualResetValueTaskSourceCore_T__GetResult)
+                    .AsMember((NamedTypeSymbol)_promiseOfValueOrEndField.Type.TypeSymbol);
 
                 // The implementation doesn't depend on the method body of the iterator method.
-                OpenMethodImplementation(IAsyncDisposable_DisposeAsync, hasMethodBodyDependency: false);
+                OpenMethodImplementation(IValueTaskSource_GetResult, hasMethodBodyDependency: false);
 
-                var bodyBuilder = ArrayBuilder<BoundStatement>.GetInstance();
+                F.CloseMethod(F.Block(
+                    // this._valueOrEndPromise.GetResult(token);
+                    F.ExpressionStatement(F.Call(F.Field(F.This(), _promiseOfValueOrEndField), promise_GetResult, F.Parameter(IValueTaskSource_GetResult.Parameters[0]))),
+                    // return;
+                    F.Return()));
+            }
+
+            // Consider factoring with IValueTaskSource<bool> implementation
+            // See https://github.com/dotnet/roslyn/issues/31517
+            private void GenerateIValueTaskSourceImplementation_GetStatus()
+            {
+                // Produce the implementation for `ValueTaskSourceStatus IValueTaskSource.GetStatus(short token)`:
+                // return this._valueOrEndPromise.GetStatus(token);
+
+                NamedTypeSymbol IValueTaskSource = F.WellKnownType(WellKnownType.System_Threading_Tasks_Sources_IValueTaskSource);
+
+                MethodSymbol IValueTaskSource_GetStatus =
+                    F.WellKnownMethod(WellKnownMember.System_Threading_Tasks_Sources_IValueTaskSource__GetStatus)
+                    .AsMember(IValueTaskSource);
```

Comment:
>.AsMember(IValueTaskSource) [](start = 20, length = 27)

Does this have any effect?

Reply:
Nope. I'll remove. Thanks

---
In reply to: [240744277](https://github.com/dotnet/roslyn/pull/31527#discussion_r240744277) [](ancestors = 240744277)